themes/bootstrap3/js/facets.js

```diff
@@ -1,87 +1,143 @@
-    facetValue.appendChild(document.createTextNode(facet.displayText));
-    item.appendChild(facetValue);
-
-    var children = null;
-    if (typeof facet.children !== 'undefined' && facet.children.length > 0) {
-      children = buildFacetNodes(facet.children, currentPath, allowExclude, excludeTitle, counts);
+
+    // Create exclude link
+    let excludeEl = null;
+    if (allowExclude && !facet.isApplied) {
+      excludeEl = el("a", "exclude");
+      excludeEl.innerHTML = VuFind.icon("facet-exclude");
+      excludeEl.setAttribute("href", currentPath + facet.exclude);
+      excludeEl.setAttribute("title", excludeTitle);
     }
 
-    json.push({
-      text: item.outerHTML,
-      children: children,
-      state: {
-        opened: facet.hasAppliedChildren,
-        selected: facet.isApplied
-      },
-      a_attr: {
-        class: 'hierarchical-facet-anchor icon-link',
-        href: url
-      },
-      data: {
-        url: url.replace(/&amp;/g, '&'),
-        count: !facet.isApplied && counts && facet.count
-          ? facet.count.toString().replace(/\B(?=(\d{3})+\b)/g, separator) : null,
-        excludeUrl: allowExclude && !facet.isApplied ? excludeUrl : '',
-        excludeTitle: excludeTitle
+    // Create facet text element
+    const orFacet = facet.operator === 'OR';
+    let textEl;
+    if (orFacet) {
+      const iconLabel = el("span", "facet-value icon-link__label");
+      iconLabel.innerText = facet.displayText;
+
+      textEl = el("span", "text");
+      textEl.innerHTML = facet.isApplied
+        ? VuFind.icon("facet-checked", { title: VuFind.translate("Selected"), class: "icon-link__icon" })
+        : VuFind.icon("facet-unchecked", "icon-link__icon");
+      textEl.append(iconLabel);
+    } else {
+      textEl = el("span", "text");
+      textEl.append(facet.displayText);
+    }
+
+    // Create link element
+    const linkEl = el("a", (orFacet ? " icon-link" : ""));
+    linkEl.setAttribute("href", currentPath + facet.href);
+    linkEl.setAttribute("title", facet.displayText);
+    linkEl.append(textEl);
+
+    // Create facet element
+    const classes = "facet js-facet-item"
+      + (facet.isApplied ? " active" : "")
+      + (orFacet ? " facetOR" : " facetAND");
+    let facetEl;
+    if (excludeEl) {
+      linkEl.className += " text";
+      facetEl = el("div", classes);
+      facetEl.append(linkEl);
+      if (badgeEl) {
+        facetEl.append(badgeEl);
       }
-    });
+      facetEl.append(excludeEl);
+    } else {
+      if (badgeEl) {
+        linkEl.append(badgeEl);
+      }
+      linkEl.className = classes + " " + linkEl.className;
+      facetEl = linkEl;
+    }
+
+    // Create toggle button
+    const toggleButton = el("button", "facet-tree__toggle-open");
+    toggleButton.setAttribute('aria-expanded', facet.hasAppliedChildren ? 'true' : 'false');
+    toggleButton.setAttribute('data-toggle-aria-expanded', '');
+    toggleButton.setAttribute('aria-label', facet.displayText);
```

Comment:
I tried to keep the changes here minimal, because #3046 will completely revamp this and we won't need this code anymore. That's why I wouldn't want to introduce a new library or templates here. Also need to keep in mind that this could be called a thousand times so any performance considerations are important as well.